[FIX] web_widget_x2many_2d_matrix: Fix linters  and enable keyboard navigation

yajo opened the pull request:
- The first commit satisfies linters, and served me as a test drive for https://github.com/OCA/pylint-odoo/pull/209 (so it satisfies those options, not current. Expect to see some linter errors in travis though).
- The 2nd commit enables keyboard navigation, and depends on https://github.com/odoo/odoo/pull/26490 for full support. Can't workaround it though, so partial support is implemented.

@Tecnativa

yajo committed:
[FIX] web_widget_x2many_2d_matrix: Enable keyboard navigation

yajo committed:
[FIX] web_widget_x2many_2d_matrix: Fix linters

pedrobaeza commented:
For better review, enable the hide whitespace changes:

![seleccion_001](https://user-images.githubusercontent.com/7165771/44450155-78aae580-a5f0-11e8-82eb-badaffca6d75.png)

simahawk commented on web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:
```diff
@@ -2,422 +2,537 @@
+                last = index === widgets.length - 1,
+                move = 0;
+            // Guess if we have to move the focus
+            if (event.data.direction === "next" && !last) {
```
can't we support also direction="down|up" to jump on the next/prev matrix line?

yajo replied:
I added it to roadmap.

simahawk reviewed (changes requested):
LG, but:
1. question about nav up/down (related to "Support cell traversal through keyboard arrows." in roadmap)
2. no version upgrade

pedrobaeza commented:
What is the status of this?

simahawk reviewed (approved):
LG

JordiBForgeFlow commented:
@mreficent can you please review?

MiquelRForgeFlow reviewed (approved):
LGTM

moylop260 commented:
@Yajo Could you fixup the commits, please?

yajo commented:
rebased.

yajo committed:
[FIX] web_widget_x2many_2d_matrix: Allow empty cells

Before this commit, if a matrix was lacking one element, some ugly errors appeared to the user.

Now, it just displays the empty cell and lets the user go on.

Also, acknowledge another limitation of the widget and add it to roadmap.

yajo commented:
Last commit also will save some cats :smile_cat: :

![captura de pantalla de 2018-08-24 11-12-57](https://user-images.githubusercontent.com/973709/44579613-da08bb00-a78e-11e8-811e-d7c082e2871c.png)

pedrobaeza commented:
:clap:

pedrobaeza commented:
I'm afraid the cats are not saved yet.

If you go in demo data to the iPad variants, disable the `Black, 16 GB` variant, and enable `Black, 32 GB`, and showing the iPad in the variant manager, you get this error:
<details>

```
TypeError: record is undefined

http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:212
Rastreo de error:
_renderLabelCell@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:212:17
_renderRow@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:189:30
_.collect@http://localhost:11069/web/static/lib/underscore/underscore.js:164:24
_renderRows@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:173:20
_renderBody@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:94:45
_renderView@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:77:25
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
_render@http://localhost:11069/web/static/src/js/views/basic/basic_renderer.js:498:16
updateState@http://localhost:11069/web/static/src/js/views/abstract_renderer.js:92:45
updateState@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/2d_matrix_renderer.js:374:20
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
_render@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/widget_x2many_2d_matrix.js:176:24
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
reset@http://localhost:11069/web/static/src/js/fields/abstract_field.js:299:16
reset@http://localhost:11069/web/static/src/js/fields/relational_fields.js:793:16
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
reset@http://localhost:11069/web/static/src/js/fields/relational_fields.js:1185:16
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
confirmChange/<@http://localhost:11069/web/static/src/js/views/basic/basic_renderer.js:110:27
_.forEach@http://localhost:11069/web/static/lib/underscore/underscore.js:145:9
confirmChange@http://localhost:11069/web/static/src/js/views/basic/basic_renderer.js:107:9
confirmChange@http://localhost:11069/web/static/src/js/views/form/form_renderer.js:108:16
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
confirmChange@http://localhost:11069/mail/static/src/js/form_renderer.js:39:16
OdooClass.extend/Class.include/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:123:35
_confirmChange@http://localhost:11069/web/static/src/js/views/basic/basic_controller.js:282:16
_applyChanges/<@http://localhost:11069/web/static/src/js/views/field_manager_mixin.js:57:28
then/</</<@http://localhost:11069/web/static/lib/jquery/jquery.js:3276:86
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
Deferred/</deferred[tuple[0]]@http://localhost:11069/web/static/lib/jquery/jquery.js:3321:33
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
Deferred/</deferred[tuple[0]]@http://localhost:11069/web/static/lib/jquery/jquery.js:3321:33
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
Deferred/</deferred[tuple[0]]@http://localhost:11069/web/static/lib/jquery/jquery.js:3321:33
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
add@http://localhost:11069/web/static/lib/jquery/jquery.js:3165:49
then/</</<@http://localhost:11069/web/static/lib/jquery/jquery.js:3278:73
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
Deferred/</deferred[tuple[0]]@http://localhost:11069/web/static/lib/jquery/jquery.js:3321:33
_applyChange/</<@http://localhost:11069/web/static/src/js/views/basic/basic_model.js:1292:25
then/</</<@http://localhost:11069/web/static/lib/jquery/jquery.js:3276:86
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
Deferred/</deferred[tuple[0]]@http://localhost:11069/web/static/lib/jquery/jquery.js:3321:33
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
add@http://localhost:11069/web/static/lib/jquery/jquery.js:3165:49
then/</</<@http://localhost:11069/web/static/lib/jquery/jquery.js:3278:73
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
Deferred/</deferred[tuple[0]]@http://localhost:11069/web/static/lib/jquery/jquery.js:3321:33
rpc/</<@http://localhost:11069/web/static/src/js/services/ajax_service.js:13:21
then/</</<@http://localhost:11069/web/static/lib/jquery/jquery.js:3276:86
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
Deferred/</deferred[tuple[0]]@http://localhost:11069/web/static/lib/jquery/jquery.js:3321:33
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
then/</</<@http://localhost:11069/web/static/lib/jquery/jquery.js:3283:73
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
then/</</<@http://localhost:11069/web/static/lib/jquery/jquery.js:3283:73
fire@http://localhost:11069/web/static/lib/jquery/jquery.js:3119:38
fireWith@http://localhost:11069/web/static/lib/jquery/jquery.js:3231:49
done@http://localhost:11069/web/static/lib/jquery/jquery.js:9275:33
callback@http://localhost:11069/web/static/lib/jquery/jquery.js:9685:57
```

</details>

pedrobaeza commented:
There's another curious thing. I'm trying this modification without the Odoo patch, and the navigation inside the widget with tab works, but when entering the widget from another widget, it gives an error. Is due to the absence of the patch? Can we workaround it?

I have recorded a video for you to see it:

![peek 24-08-2018 21-56](https://user-images.githubusercontent.com/7165771/44605242-e15cb280-a7e8-11e8-9a6a-c1d60d47d1c2.gif)

<details>

```
TypeError: options.event is undefined

http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/widget_x2many_2d_matrix.js:198
Rastreo de error:
activate@http://localhost:11069/web_widget_x2many_2d_matrix/static/src/js/widget_x2many_2d_matrix.js:198:13
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
_activateFieldWidget@http://localhost:11069/web/static/src/js/views/basic/basic_renderer.js:187:29
_activateNextFieldWidget@http://localhost:11069/web/static/src/js/views/basic/basic_renderer.js:220:16
_onNavigationMove@http://localhost:11069/web/static/src/js/views/form/form_renderer.js:882:13
proxy/<@http://localhost:11069/web/static/src/js/core/mixins.js:279:20
trigger@http://localhost:11069/web/static/src/js/core/mixins.js:226:17
_trigger_up@http://localhost:11069/web/static/src/js/core/mixins.js:339:9
_trigger_up@http://localhost:11069/web/static/src/js/core/mixins.js:341:13
trigger_up@http://localhost:11069/web/static/src/js/core/mixins.js:335:9
_onKeydown@http://localhost:11069/web/static/src/js/fields/abstract_field.js:464:17
_onKeydown@http://localhost:11069/web/static/src/js/fields/relational_fields.js:601:9
OdooClass.extend/</prototype[name]</<@http://localhost:11069/web/static/src/js/core/class.js:90:35
proxy/<@http://localhost:11069/web/static/src/js/core/mixins.js:279:20
dispatch@http://localhost:11069/web/static/lib/jquery/jquery.js:4640:50
add/elemData.handle@http://localhost:11069/web/static/lib/jquery/jquery.js:4309:41
```

</details>

yajo commented:
Check it now please

yajo commented:
Finally managed to reproduce and fix https://github.com/OCA/web/pull/1022#issuecomment-415863235. Can you review please?

pedrobaeza commented:
Thanks for doing the dirty work. Merging!
